openmmtools/alchemy.py

```diff
@@ -12,413 +12,905 @@
-        [self.alchemically_modified_system, self.force_labels] = self._createAlchemicallyModifiedSystem(self.reference_system)
-
-        # Build a list of all alchemical parameters available in this system.
-        self.alchemical_parameters = self._getSystemGlobalParameters(self.alchemically_modified_system, prefix='lambda_')
-
-        # Store information for use in aiding debugging of alchemical factory
-        self.test_positions = test_positions
-        self.platform = platform
-        if self.test_positions is not None:
-            self._checkEnergyIsFinite(self.alchemically_modified_system, test_positions, platform=platform)
-
-        # DEBUG: Write XML
-        debug_write_xml = False
-        if debug_write_xml:
-            import os, os.path
-            def write_file(filename, contents):
-                with open(filename, 'w') as outfile:
-                    outfile.write(contents)
-
-            logger.info("Serializing to XML...")
-            system_filename = os.path.join('setup', 'alchemical-system.xml')
-            write_file(system_filename, openmm.XmlSerializer.serialize(self.alchemically_modified_system))
-
-        return
+        # Add particles.
+        for particle_index in range(reference_system.getNumParticles()):
+            mass = reference_system.getParticleMass(particle_index)
+            alchemical_system.addParticle(mass)
+
+        # Add constraints.
+        for constraint_index in range(reference_system.getNumConstraints()):
+            atom_i, atom_j, r0 = reference_system.getConstraintParameters(constraint_index)
+            alchemical_system.addConstraint(atom_i, atom_j, r0)
+
+        # Modify forces as appropriate, copying other forces without modification.
+        for reference_force in reference_system.getForces():
+            # TODO switch to functools.singledispatch when drop Python2
+            reference_force_name = reference_force.__class__.__name__
+            alchemical_force_creator_name = '_alchemically_modify_{}'.format(reference_force_name)
+            try:
+                alchemical_force_creator_func = getattr(self, alchemical_force_creator_name)
+            except AttributeError:
+                alchemical_forces = [copy.deepcopy(reference_force)]
+            else:
+                alchemical_forces = alchemical_force_creator_func(reference_force, alchemical_region)
+            for alchemical_force in alchemical_forces:
+                alchemical_system.addForce(alchemical_force)
+
+        # Record timing statistics.
+        timer.stop('Create alchemically modified system')
+        timer.report_timing()
+
+        return alchemical_system
 
     @classmethod
-    def _getSystemGlobalParameters(cls, system, prefix='lambda_'):
-        """
-        Get a list of available alchemical parameters beginning with 'prefix'
+    def get_energy_components(cls, alchemical_system, alchemical_state, positions, platform=None):
```

Comment:
We use this only in tests currently. It may be useful to debug if we are hunting down problems, but this won't be used normally.